experimental/printer/dom/chunk.go

```diff
@@ -0,0 +1,229 @@
+func (c *Chunk) SplitKind() SplitKind {
+	if !c.hasText {
+		panic("protocompile/printer/dom: called SplitKind() on unset Chunk")
+	}
+	return c.splitKind
+}
+
+// SetSplitKind sets the SplitKind of the Chunk.
+func (c *Chunk) SetSplitKind(splitKind SplitKind) {
+	c.splitKind = splitKind
+}
+
+// SetChild sets the Chunk's child.
+func (c *Chunk) SetChild(child *Chunks) {
+	if child != nil && len(child.chunks) > 0 {
+		c.child = child
+	}
+}
+
+// SetIndentDepth sets the indent depth of the Chunk.
+func (c *Chunk) SetIndentDepth(indentDepth int) {
+	c.indentDepth = indentDepth
+}
+
+// SetSpaceIfUnsplit sets whether the Chunk will output a space if it is not split.
+func (c *Chunk) SetSpaceIfUnsplit(spaceIfUnsplit bool) {
+	c.spaceIfUnsplit = spaceIfUnsplit
+}
+
+// SetSplitKindIfSplit sets the SplitKind of the Chunk if it is split. This will panic if it
+// is not called with SplitKindHard or SplitKindDouble.
+func (c *Chunk) SetSplitKindIfSplit(splitKindIfSplit SplitKind) {
+	if splitKindIfSplit != SplitKindHard && splitKindIfSplit != SplitKindDouble {
+		panic(fmt.Sprintf(
+			"protocompile/printer/dom: called SetSplitKindIfSplit with %s, this is restricted to %s and %s",
+			splitKindIfSplit,
+			SplitKindHard,
+			SplitKindDouble,
+		))
+	}
+	c.splitKindIfSplit = splitKindIfSplit
+}
+
+// The Chunk's current text.
+//
+// Unformatted chunks will return the raw text set on the chunk.
+// Formatted chunks should rely on the indenting and splitting logic set on the chunk, so
+// this will return the text with leading and trailing whitespaces trimmed.
+func (c *Chunk) text() string {
+	if c.formatted {
+		return strings.TrimSpace(c.rawText)
+	}
+	return c.rawText
+}
+
+// The Chunk's indent string.
+//
+// Unformatted chunks will always return empty string.
+// Formatted chunks will return an indent string only if the chunk is indented and only
+// if it is not a whitespace-only chunk. The indent is based on the indent depth, and indent
+// string set on the chunk..
+func (c *Chunk) indent() string {
+	if c.formatted && c.indented && !c.whitespaceOnly {
+		return strings.Repeat(c.indentStr, c.indentDepth)
+	}
+	return ""
+}
+
+// The Chunk's split string (trailing whitespace).
+//
+// Unformatted chunks will always return empty string.
+// Formatted chunks will return a split string only if the chunk is not whitespace-only.
+// The split string is based on the split kind currently set and space if unsplit for
+// soft/never split chunks.
+func (c *Chunk) splitString() string {
+	if c.formatted && !c.whitespaceOnly {
+		switch c.splitKind {
+		case SplitKindHard:
+			return "\n"
+		case SplitKindDouble:
+			return "\n\n"
+		case SplitKindSoft, SplitKindNever:
+			if c.spaceIfUnsplit {
+				return space
+			}
+		}
+	}
+	return ""
+}
+
+// The Chunk's output string.
+//
+// TODO: should this be named "render", since this is technically rendering the string
+// output of the Chunk?
```

Comment:
Small naming thing... I'm open to either, although after writing some docs, I might have a preference for `render`.